igvm/commands.py

```diff
@@ -145,10 +158,113 @@ def vm_build(vm_hostname, localimage=None, nopuppet=False, postboot=None,
     )
 
 
+@with_fabric_settings   # NOQA: C901
+def vm_migrate(vm_hostname, hypervisor_hostname=None, newip=None,
+               runpuppet=False, maintenance=False, offline=False,
+               ignore_reserved=False):
+    """Migrate a VM to a new hypervisor."""
+    vm = _get_vm(vm_hostname, ignore_reserved=ignore_reserved)
+
+    # If not specified automatically find a new better hypervisor
+    if hypervisor_hostname:
+        hypervisor = _get_hypervisor(
+            hypervisor_hostname, ignore_reserved=ignore_reserved
+        )
+    else:
+        hypervisor = vm.get_best_hypervisor(
+            ['online', 'online_reserved'] if ignore_reserved else ['online']
+        )
+
+    was_running = vm.is_running()
+
+    # There is no point of online migration, if the VM is already shutdown.
+    if not was_running:
+        offline = True
+
+    if not offline and newip:
+        raise IGVMError('Online migration cannot change IP address.')
+
+    if not offline and runpuppet:
+        raise IGVMError('Online migration cannot run Puppet.')
+
+    if not runpuppet and newip:
```

Comment:
Out of curiosity, whats the reason to run puppet but not change the IP on migration ?

Reply:
Maybe @TuxPowered42 knows?

Reply:
I don't know. It was always an option to run puppet after migration. I don't think there is any other use for it than to change IP address of a VM.